QueryReactiveApi The exception could not be delivered to the consumer because it has already canceled/disposed

<!--

Thank you for reporting a bug. 

* Please add a :+1: or comment on a similar existing bug report instead of opening a new one.
    * https://github.com/influxdata/influxdb-client-java/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+is%3Aclosed+sort%3Aupdated-desc+label%3Abug+
* Please check whether the bug can be reproduced with the latest release.
* The fastest way to fix a bug is to open a Pull Request.
    * https://github.com/influxdata/influxdb-client-java/pulls

-->

__Steps to reproduce:__
List the minimal actions needed to reproduce the behavior.

1. Create a long running request with `QueryReactiveApi` and schedule it as a `Future`
2. Cancel the future forcibly (`future.cancel(true);`)

```java

QueryReactiveApi queryApi = client.getQueryReactiveApi();

Disposable disposable;
Future<?> future = executor.scheduleWithFixedDelay(() -> {
    disposable = queryApi.queryRaw(query)
        .doOnNext(onNext -> {
            System.out.println("onNext");
        }).subscribe();
}, 0, 60, TimeUnit.SECONDS);

future.cancel(true);
```

__Expected behavior:__
The request would be canceled silently

__Actual behavior:__
Huge pile of stack traces:

```
io.reactivex.exceptions.UndeliverableException: The exception could not be delivered to the consumer because it has already canceled/disposed the flow or the exception has nowhere to go to begin with. Further reading: https://github.com/ReactiveX/RxJava/wiki/What's-different-in-2.0#error-handling | java.io.InterruptedIOException: interrupted
        at io.reactivex.plugins.RxJavaPlugins.onError(RxJavaPlugins.java:367)
        at io.reactivex.internal.operators.observable.ObservableCreate$CreateEmitter.onError(ObservableCreate.java:73)
        at com.influxdb.internal.AbstractRestClient.catchOrPropagateException(AbstractRestClient.java:158)
        at com.influxdb.internal.AbstractQueryApi.lambda$queryRaw$4(AbstractQueryApi.java:140)
        at com.influxdb.internal.AbstractQueryApi.lambda$query$5(AbstractQueryApi.java:174)
        at com.influxdb.internal.AbstractQueryApi$1.onResponse(AbstractQueryApi.java:220)
        at com.influxdb.internal.AbstractQueryApi.query(AbstractQueryApi.java:238)
        at com.influxdb.internal.AbstractQueryApi.query(AbstractQueryApi.java:190)
        at com.influxdb.internal.AbstractQueryApi.queryRaw(AbstractQueryApi.java:144)
        at com.influxdb.client.reactive.internal.QueryReactiveApiImpl.lambda$null$8(QueryReactiveApiImpl.java:334)
        at io.reactivex.internal.operators.observable.ObservableCreate.subscribeActual(ObservableCreate.java:40)
        at io.reactivex.Observable.subscribe(Observable.java:12284)
        at io.reactivex.internal.operators.flowable.FlowableFromObservable.subscribeActual(FlowableFromObservable.java:29)
        at io.reactivex.Flowable.subscribe(Flowable.java:14935)
        at io.reactivex.internal.operators.flowable.FlowableOnBackpressureBuffer.subscribeActual(FlowableOnBackpressureBuffer.java:46)
        at io.reactivex.Flowable.subscribe(Flowable.java:14935)
        at io.reactivex.Flowable.subscribe(Flowable.java:14882)
        at io.reactivex.internal.operators.flowable.FlowableFlatMap$MergeSubscriber.onNext(FlowableFlatMap.java:163)
        at io.reactivex.internal.operators.flowable.FlowableMap$MapSubscriber.onNext(FlowableMap.java:68)
        at io.reactivex.internal.operators.flowable.FlowableMap$MapSubscriber.onNext(FlowableMap.java:68)
        at io.reactivex.internal.subscriptions.ScalarSubscription.request(ScalarSubscription.java:55)
        at io.reactivex.internal.subscribers.BasicFuseableSubscriber.request(BasicFuseableSubscriber.java:153)
        at io.reactivex.internal.subscribers.BasicFuseableSubscriber.request(BasicFuseableSubscriber.java:153)
        at io.reactivex.internal.operators.flowable.FlowableFlatMap$MergeSubscriber.onSubscribe(FlowableFlatMap.java:117)
        at io.reactivex.internal.subscribers.BasicFuseableSubscriber.onSubscribe(BasicFuseableSubscriber.java:67)
        at io.reactivex.internal.subscribers.BasicFuseableSubscriber.onSubscribe(BasicFuseableSubscriber.java:67)
        at io.reactivex.internal.operators.flowable.FlowableJust.subscribeActual(FlowableJust.java:34)
        at io.reactivex.Flowable.subscribe(Flowable.java:14935)
        at io.reactivex.internal.operators.flowable.FlowableMap.subscribeActual(FlowableMap.java:37)
        at io.reactivex.Flowable.subscribe(Flowable.java:14935)
        at io.reactivex.internal.operators.flowable.FlowableMap.subscribeActual(FlowableMap.java:37)
        at io.reactivex.Flowable.subscribe(Flowable.java:14935)
        at io.reactivex.internal.operators.flowable.FlowableFlatMap.subscribeActual(FlowableFlatMap.java:53)
        at io.reactivex.Flowable.subscribe(Flowable.java:14935)
        at io.reactivex.internal.operators.flowable.FlowableDoOnEach.subscribeActual(FlowableDoOnEach.java:50)
        at io.reactivex.Flowable.subscribe(Flowable.java:14935)
        at io.reactivex.internal.operators.flowable.FlowableDoOnEach.subscribeActual(FlowableDoOnEach.java:50)
        at io.reactivex.Flowable.subscribe(Flowable.java:14935)
        at io.reactivex.internal.operators.flowable.FlowableDoOnEach.subscribeActual(FlowableDoOnEach.java:50)
        at io.reactivex.Flowable.subscribe(Flowable.java:14935)
        at io.reactivex.Flowable.subscribe(Flowable.java:14872)
        at io.reactivex.Flowable.subscribe(Flowable.java:14730)
        at main.Main.main(Main.java:33)
Caused by: java.io.InterruptedIOException: interrupted
        at okio.Timeout.throwIfReached(Timeout.kt:98)
        at okio.InputStreamSource.read(JvmOkio.kt:87)
        at okio.AsyncTimeout$source$1.read(AsyncTimeout.kt:129)
        at okio.RealBufferedSource.read(RealBufferedSource.kt:188)
        at okhttp3.internal.http1.Http1ExchangeCodec$AbstractSource.read(Http1ExchangeCodec.java:389)
        at okhttp3.internal.http1.Http1ExchangeCodec$ChunkedSource.read(Http1ExchangeCodec.java:475)
        at okhttp3.internal.connection.Exchange$ResponseBodySource.read(Exchange.java:286)
        at okio.RealBufferedSource.read(RealBufferedSource.kt:188)
        at okio.RealBufferedSource.exhausted(RealBufferedSource.kt:198)
        at okio.InflaterSource.refill(InflaterSource.kt:112)
        at okio.InflaterSource.readOrInflate(InflaterSource.kt:76)
        at okio.InflaterSource.read(InflaterSource.kt:49)
        at okio.GzipSource.read(GzipSource.kt:69)
        at okio.RealBufferedSource.read(RealBufferedSource.kt:188)
        at okio.ForwardingSource.read(ForwardingSource.kt:29)
        at retrofit2.OkHttpCall$ExceptionCatchingResponseBody$1.read(OkHttpCall.java:314)
        at okio.RealBufferedSource.indexOf(RealBufferedSource.kt:449)
        at okio.RealBufferedSource.indexOf(RealBufferedSource.kt:117)
        at okio.RealBufferedSource.readUtf8Line(RealBufferedSource.kt:319)
        at com.influxdb.internal.AbstractQueryApi.parseFluxResponseToLines(AbstractQueryApi.java:249)
        at com.influxdb.internal.AbstractQueryApi.lambda$queryRaw$4(AbstractQueryApi.java:138)
        ... 47 more
```

Shouldn't `AbstractRestClient` check if the onError-handler is disposed before calling `accept(exception)`:

```java
    void catchOrPropagateException(@Nonnull final Exception exception,
                                   @Nonnull final Consumer<? super Throwable> onError) {

        Arguments.checkNotNull(exception, "exception");
        Arguments.checkNotNull(onError, "onError");

        //
        // Socket closed by remote server or end of data
        //
        if (isCloseException(exception)) {
            LOG.log(Level.FINEST, "Socket closed by remote server or end of data", exception);
        } else {
            // here check if onError is disposed already ?
            onError.accept(exception);
        }
    }
```

Or alternatively add a case to the if-clause and check for interruptions?

__Specifications:__
 - Client Version: 4.3.0
 - InfluxDB Version: 2.1.1
 - JDK Version: 11
 - Platform: Linux

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,8 @@
 ## 5.1.0 [unreleased]
 
+### Bug Fixes
+1. [#313](https://github.com/influxdata/influxdb-client-java/pull/313): Do not deliver `exception` when the consumer is already disposed [influxdb-client-reactive]
+
 ## 5.0.0 [2022-03-18]
 
 ### Breaking Changes
@@ -12,7 +15,6 @@
 
 ### Bug Fixes
 1. [#303](https://github.com/influxdata/influxdb-client-java/pull/303): Change `PermissionResource.type` to `String`
-1. [#313](https://github.com/influxdata/influxdb-client-java/pull/313): Do not deliver `exception` when the consumer is already disposed [influxdb-client-reactive]
 
 ### CI
 1. [#304](https://github.com/influxdata/influxdb-client-java/pull/304): Use new Codecov uploader for reporting code coverage

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -12,6 +12,7 @@
 
 ### Bug Fixes
 1. [#303](https://github.com/influxdata/influxdb-client-java/pull/303): Change `PermissionResource.type` to `String`
+1. [#313](https://github.com/influxdata/influxdb-client-java/pull/313): Do not deliver `exception` when the consumer is already disposed [influxdb-client-reactive]
 
 ### CI
 1. [#304](https://github.com/influxdata/influxdb-client-java/pull/304): Use new Codecov uploader for reporting code coverage

--- a/client-core/src/main/java/com/influxdb/internal/AbstractQueryApi.java
+++ b/client-core/src/main/java/com/influxdb/internal/AbstractQueryApi.java
@@ -62,7 +62,7 @@ import retrofit2.Response;
  */
 public abstract class AbstractQueryApi extends AbstractRestClient {
 
-    private static final Logger LOG = Logger.getLogger(AbstractQueryApi.class.getName());
+    protected static final Logger LOG = Logger.getLogger(AbstractQueryApi.class.getName());
 
     protected final FluxCsvParser fluxCsvParser = new FluxCsvParser();
     protected final FluxResultMapper resultMapper = new FluxResultMapper();

--- a/client-reactive/src/main/java/com/influxdb/client/reactive/internal/QueryReactiveApiImpl.java
+++ b/client-reactive/src/main/java/com/influxdb/client/reactive/internal/QueryReactiveApiImpl.java
@@ -22,6 +22,8 @@
 package com.influxdb.client.reactive.internal;
 
 import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.logging.Level;
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
 
@@ -41,6 +43,8 @@ import com.influxdb.utils.Arguments;
 import io.reactivex.BackpressureStrategy;
 import io.reactivex.Flowable;
 import io.reactivex.Observable;
+import io.reactivex.ObservableEmitter;
+import org.jetbrains.annotations.NotNull;
 import org.reactivestreams.Publisher;
 
 /**
@@ -187,7 +191,7 @@ final class QueryReactiveApiImpl extends AbstractQueryApi implements QueryReacti
                             }
                         };
 
-                        query(queryCall, consumer, subscriber::onError, subscriber::onComplete, false);
+                        query(queryCall, consumer, onError(subscriber), subscriber::onComplete, false);
                     });
 
                     return observable.toFlowable(BackpressureStrategy.BUFFER);
@@ -331,7 +335,7 @@ final class QueryReactiveApiImpl extends AbstractQueryApi implements QueryReacti
                         }
                     };
 
-                    queryRaw(queryCall, consumer, subscriber::onError, subscriber::onComplete, false);
+                    queryRaw(queryCall, consumer, onError(subscriber), subscriber::onComplete, false);
                 });
 
                 return observable.toFlowable(BackpressureStrategy.BUFFER);
@@ -349,5 +353,17 @@ final class QueryReactiveApiImpl extends AbstractQueryApi implements QueryReacti
 
         return queryRawQuery(Flowable.fromPublisher(queryStream)
             .map(q -> new Query().query(q).dialect(dialect)), dialect, org);
+    }
+
+    @NotNull
+    private Consumer<Throwable> onError(final ObservableEmitter<?> subscriber) {
+        return throwable -> {
+            if (!subscriber.isDisposed()) {
+                subscriber.onError(throwable);
+            } else {
+                LOG.log(Level.FINEST, "The exception could not be delivered to the consumer "
+                        + "because it has already canceled/disposed.", throwable);
+            }
+        };
     }
 }

--- a/client-reactive/src/test/java/com/influxdb/client/reactive/QueryReactiveApiTest.java
+++ b/client-reactive/src/test/java/com/influxdb/client/reactive/QueryReactiveApiTest.java
@@ -1,0 +1,86 @@
+/*
+ * The MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package com.influxdb.client.reactive;
+
+import java.io.InterruptedIOException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+
+import com.influxdb.test.AbstractMockServerTest;
+
+import io.reactivex.Flowable;
+import org.assertj.core.api.Assertions;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.platform.runner.JUnitPlatform;
+import org.junit.runner.RunWith;
+
+/**
+ * @author Jakub Bednar (03/16/2022 08:28)
+ */
+@RunWith(JUnitPlatform.class)
+class QueryReactiveApiTest extends AbstractMockServerTest {
+
+    private InfluxDBClientReactive influxDBClient;
+
+    @BeforeEach
+    void setUp() {
+
+        influxDBClient = InfluxDBClientReactiveFactory.create(startMockServer());
+    }
+
+    @AfterEach
+    void tearDown() {
+
+        influxDBClient.close();
+    }
+
+    @Test
+    public void doNotPropagateErrorOnCanceledConsumer() throws InterruptedException {
+
+        mockServer.enqueue(createErrorResponse("Request Timeout", true, 408)
+                .setBodyDelay(3, TimeUnit.SECONDS));
+
+        QueryReactiveApi queryApi = influxDBClient.getQueryReactiveApi();
+
+        final Throwable[] throwable = new Throwable[1];
+        Future<?> future = Executors
+                .newSingleThreadScheduledExecutor()
+                .scheduleWithFixedDelay(() -> Flowable
+                                .fromPublisher(queryApi.queryRaw("from()", "my-org"))
+                                .subscribe(
+                                        s -> {
+
+                                        },
+                                        t -> throwable[0] = t),
+                        0, 1, TimeUnit.SECONDS);
+
+        Thread.sleep(2_000);
+        future.cancel(true);
+        Thread.sleep(2_000);
+
+        Assertions.assertThat(throwable[0]).isNotNull();
+        Assertions.assertThat(throwable[0]).isInstanceOf(InterruptedIOException.class);
+    }
+}

